xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+        self._meta: dict[str, Any] = {"bias_func": bias_func}
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optional[rio.transform.Affine],
+                  weights: Optional[np.ndarray]):
+        """Estimate the bias using the bias_func."""
+        diff = ref_dem - tba_dem
+        diff = diff[np.isfinite(diff)]
+
+        # Use weights if those were provided.
+        bias = self._meta["bias_func"](diff) if weights is None \
+            else self._meta["bias_func"](diff, weights=weights)
+
+        self._meta["bias"] = bias
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the bias to a DEM."""
+        return dem + self._meta["bias"]
+
+    def _apply_pts_func(self, coords: np.ndarray):
+        """Apply the bias to a given coordinate array."""
+        new_coords = coords.copy()
+        new_coords[:, 2] += self._apply_func(coords[:, 2], None)  # type: ignore
+
+        return new_coords
+
+    def _to_matrix_func(self) -> np.ndarray:
+        """Convert the bias to a transform matrix."""
+        empty_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]], dtype=float)
+
+        empty_matrix[2, 3] += self._meta["bias"]
+
+        return empty_matrix
+
+
+class ICP(Coreg):
+    """
+    Iterative Closest Point DEM coregistration.
+
+    Estimates a rigid transform (rotation + translation) between two DEMs.
+
+    Requires 'opencv'
+    """
+
+    def __init__(self, max_iterations=100, tolerance=0.05, rejection_scale=2.5, num_levels=6):  # pylint: disable=super-init-not-called
+        """
+        Instantiate an ICP coregistration object.
+
+        :param max_iterations: The maximum allowed iterations before stopping.
+        :param tolerance: The residual change threshold after which to stop the iterations.
+        :param rejection_scale: The threshold (std * rejection_scale) to consider points as outliers.
+        :param num_levels: Number of octree levels to consider. A higher number is faster but may be more inaccurate.
+        """
+        if not _has_cv2:
+            raise ValueError("Optional dependency needed. Install 'opencv'")
+        self.max_iterations = max_iterations
+        self.tolerance = tolerance
+        self.rejection_scale = rejection_scale
+        self.num_levels = num_levels
+        self._meta: dict[str, Any] = {}
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optional[rio.transform.Affine],
+                  weights: Optional[np.ndarray]):
+        """Estimate the rigid transform from tba_dem to ref_dem."""
+        if weights is not None:
+            warnings.warn("ICP was given weights, but does not support it.")
+
+        bounds, resolution = _transform_to_bounds_and_res(ref_dem, transform)
+        points: dict[str, np.ndarray] = {}
+        # Generate the x and y coordinates for the reference_dem
+        x_coords, y_coords = np.meshgrid(
```

Comment:
See my comment about subsampling. It could be done at this stage for example, to reduce memory usage.